Error with channel slider due to #276

I have found another problem that appeared after #276 (this other problem, https://github.com/scverse/napari-spatialdata/issues/282 is now fixed).

I'll open a PR with tests to reproduce the problem.

This is the error that I get (to reproduce please see the code in the attached PR). Interestingly the problem is only for single scale images. Also, it seems that it is due to the presence of coordinate transformations, since the values of current_points appear to operate in transformed coordinates, while the tensors can be smaller.

Traceback:
```
FAILED [100%]2024-08-21 16:44:43.396 | INFO     | napari_spatialdata._view:_on_layer_update:347 - Updating layer.

tests/test_view.py:40 (test_channel_slider_images_bug_302[QtAdataViewWidget-False])
qtbot = <pytestqt.qtbot.QtBot object at 0x172b525c0>
make_napari_viewer = <function make_napari_viewer.<locals>.actual_factory at 0x3177d45e0>
widget = <class 'napari_spatialdata._view.QtAdataViewWidget'>
multiscale = False

    @pytest.mark.parametrize("multiscale", [True, False])
    @pytest.mark.parametrize("widget", [QtAdataViewWidget])
    def test_channel_slider_images_bug_302(qtbot, make_napari_viewer: any, widget: Any, multiscale: bool):
        # https://github.com/scverse/napari-spatialdata/issues/302
        channels = [f"channel_{i}" for i in range(3)]
        sdata_blobs = blobs(c_coords=channels)
    
        # set a transformation to the single-scale and multi-scale images
        element_name = "blobs_image" if not multiscale else "blobs_multiscale_image"
        affine = Affine(np.array([[10, 20, 30], [40, 50, 60], [0, 0, 1]]), input_axes=("x", "y"), output_axes=("x", "y"))
        set_transformation(sdata_blobs[element_name], transformation=affine, to_coordinate_system="global")
    
        viewer = make_napari_viewer()
        sdata_widget = SdataWidget(viewer, EventedList([sdata_blobs]))
    
        viewer.window.add_dock_widget(sdata_widget, name="SpatialData")
        sdata_widget.viewer_model.add_sdata_image(sdata_blobs, element_name, "global", False)
    
        # this connects the slider to the viewer (done in __init__)
        _ = widget(viewer)
    
        # check if the slider is present
        start, stop, step = viewer.dims.range[0]
        assert start == 0
        assert step == 1
    
        # simulate position change of the slider
>       viewer.dims.set_current_step(0, 0)

test_view.py:68: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/napari/components/dims.py:361: in set_current_step
    self.set_point(axis, value_world)
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/napari/components/dims.py:346: in set_point
    self.point = tuple(full_point)
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/napari/utils/events/evented_model.py:307: in __setattr__
    with ComparisonDelayer(self):
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/napari/utils/events/evented_model.py:520: in __exit__
    self._target._check_if_values_changed_and_emit_if_needed()
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/napari/utils/events/evented_model.py:345: in _check_if_values_changed_and_emit_if_needed
    getattr(self.events, name)(value=new_value)
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/napari/utils/events/event.py:764: in __call__
    self._invoke_callback(cb, event if pass_event else None)
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/napari/utils/events/event.py:802: in _invoke_callback
    _handle_exception(
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/napari/utils/events/event.py:789: in _invoke_callback
    cb(event)
../src/napari_spatialdata/_view.py:323: in _channel_changed
    image = layer.data[tuple(current_point)].compute()
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/xarray/core/dataarray.py:891: in __getitem__
    return self.isel(indexers=self._item_key_to_dict(key))
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/xarray/core/dataarray.py:1527: in isel
    variable = self._variable.isel(indexers, missing_dims=missing_dims)
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/xarray/core/variable.py:1034: in isel
    return self[key]
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/xarray/core/variable.py:801: in __getitem__
    data = indexing.apply_indexer(indexable, indexer)
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/xarray/core/indexing.py:1030: in apply_indexer
    return indexable[indexer]
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/xarray/core/indexing.py:1634: in __getitem__
    return self.array[indexer.tuple]
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/dask/array/core.py:1975: in __getitem__
    index2 = normalize_index(index, self.shape)
/opt/miniconda3/envs/ome/lib/python3.10/site-packages/dask/array/slicing.py:802: in normalize_index
    check_index(axis, i, d)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

axis = 2, ind = 1375, dimension = 512

    def check_index(axis, ind, dimension):
        """Check validity of index for a given dimension
    
        Examples
        --------
        >>> check_index(0, 3, 5)
        >>> check_index(0, 5, 5)
        Traceback (most recent call last):
        ...
        IndexError: Index 5 is out of bounds for axis 0 with size 5
    
        >>> check_index(1, 6, 5)
        Traceback (most recent call last):
        ...
        IndexError: Index 6 is out of bounds for axis 1 with size 5
    
        >>> check_index(1, -1, 5)
        >>> check_index(1, -6, 5)
        Traceback (most recent call last):
        ...
        IndexError: Index -6 is out of bounds for axis 1 with size 5
    
        >>> check_index(0, [1, 2], 5)
        >>> check_index(0, [6, 3], 5)
        Traceback (most recent call last):
        ...
        IndexError: Index is out of bounds for axis 0 with size 5
    
        >>> check_index(1, slice(0, 3), 5)
    
        >>> check_index(0, [True], 1)
        >>> check_index(0, [True, True], 3)
        Traceback (most recent call last):
        ...
        IndexError: Boolean array with size 2 is not long enough for axis 0 with size 3
        >>> check_index(0, [True, True, True], 1)
        Traceback (most recent call last):
        ...
        IndexError: Boolean array with size 3 is not long enough for axis 0 with size 1
        """
        if isinstance(ind, list):
            ind = np.asanyarray(ind)
    
        # unknown dimension, assumed to be in bounds
        if np.isnan(dimension):
            return
        elif is_dask_collection(ind):
            return
        elif is_arraylike(ind):
            if ind.dtype == bool:
                if ind.size != dimension:
                    raise IndexError(
                        f"Boolean array with size {ind.size} is not long enough "
                        f"for axis {axis} with size {dimension}"
                    )
            elif (ind >= dimension).any() or (ind < -dimension).any():
                raise IndexError(
                    f"Index is out of bounds for axis {axis} with size {dimension}"
                )
        elif isinstance(ind, slice):
            return
        elif ind is None:
            return
    
        elif ind >= dimension or ind < -dimension:
>           raise IndexError(
                f"Index {ind} is out of bounds for axis {axis} with size {dimension}"
            )
E           IndexError: Index 1375 is out of bounds for axis 2 with size 512

/opt/miniconda3/envs/ome/lib/python3.10/site-packages/dask/array/slicing.py:874: IndexError
```

--- a/src/napari_spatialdata/_view.py
+++ b/src/napari_spatialdata/_view.py
@@ -307,12 +307,16 @@ class QtAdataViewWidget(QWidget):
 
         current_point = list(event.value)
         displayed = self._viewer.dims.displayed
-        for i, (lo_size, hi_size, cord) in enumerate(zip(layer.data[-1].shape, layer.data[0].shape, current_point)):
-            if i in displayed:
-                current_point[i] = slice(None)
-            else:
-                current_point[i] = int(cord * lo_size / hi_size)
-
+        if layer.multiscale:
+            for i, (lo_size, hi_size, cord) in enumerate(zip(layer.data[-1].shape, layer.data[0].shape, current_point)):
+                if i in displayed:
+                    current_point[i] = slice(None)
+                else:
+                    current_point[i] = int(cord * lo_size / hi_size)
+        else:
+            for i in range(len(current_point)):
+                if i in displayed:
+                    current_point[i] = slice(None)
         # TODO remove once contrast limits in napari are fixed
         if isinstance(layer.data, MultiScaleData):
             # just compute lowest resolution

--- a/tests/test_view.py
+++ b/tests/test_view.py
@@ -1,15 +1,18 @@
 from typing import Any
 
+import numpy as np
 import pytest
 from napari.utils.events import EventedList
 from napari_spatialdata._sdata_widgets import SdataWidget
 from napari_spatialdata._view import QtAdataViewWidget
 from spatialdata.datasets import blobs
+from spatialdata.transformations import Affine, set_transformation
 
 
 @pytest.mark.parametrize("widget", [QtAdataViewWidget])
 @pytest.mark.parametrize("n_channels", [3, 5])
-def test_channel_slider_images(qtbot, make_napari_viewer: any, widget: Any, n_channels: int):
+def test_channel_slider_images_bug_282(qtbot, make_napari_viewer: any, widget: Any, n_channels: int):
+    # https://github.com/scverse/napari-spatialdata/issues/282
     channels = [f"channel_{i}" for i in range(n_channels)]
     sdata_blobs = blobs(c_coords=channels)
     viewer = make_napari_viewer()
@@ -17,6 +20,41 @@ def test_channel_slider_images(qtbot, make_napari_viewer: any, widget: Any, n_ch
 
     viewer.window.add_dock_widget(sdata_widget, name="SpatialData")
     sdata_widget.viewer_model.add_sdata_image(sdata_blobs, "blobs_image", "global", False)
+
+    # this connects the slider to the viewer (done in __init__)
+    _ = widget(viewer)
+
+    # check if the slider is present
+    start, stop, step = viewer.dims.range[0]
+    assert start == 0
+    assert step == 1
+
+    # simulate position change of the slider
+    viewer.dims.set_current_step(0, 0)
+    qtbot.wait(50)  # wait for a short time to simulate user interaction
+    viewer.dims.set_current_step(0, 1)
+    qtbot.wait(50)  # wait for a short time to simulate user interaction
+
+    viewer.close()
+
+
+@pytest.mark.parametrize("multiscale", [True, False])
+@pytest.mark.parametrize("widget", [QtAdataViewWidget])
+def test_channel_slider_images_bug_302(qtbot, make_napari_viewer: any, widget: Any, multiscale: bool):
+    # https://github.com/scverse/napari-spatialdata/issues/302
+    channels = [f"channel_{i}" for i in range(3)]
+    sdata_blobs = blobs(c_coords=channels)
+
+    # set a transformation to the single-scale and multi-scale images
+    element_name = "blobs_image" if not multiscale else "blobs_multiscale_image"
+    affine = Affine(np.array([[10, 20, 30], [40, 50, 60], [0, 0, 1]]), input_axes=("x", "y"), output_axes=("x", "y"))
+    set_transformation(sdata_blobs[element_name], transformation=affine, to_coordinate_system="global")
+
+    viewer = make_napari_viewer()
+    sdata_widget = SdataWidget(viewer, EventedList([sdata_blobs]))
+
+    viewer.window.add_dock_widget(sdata_widget, name="SpatialData")
+    sdata_widget.viewer_model.add_sdata_image(sdata_blobs, element_name, "global", False)
 
     # this connects the slider to the viewer (done in __init__)
     _ = widget(viewer)

